include/libopencm3/usb/usbd.h

```diff
@@ -102,18 +102,40 @@ extern void usbd_register_resume_callback(usbd_device *usbd_dev,
 extern void usbd_register_sof_callback(usbd_device *usbd_dev,
 				       void (*callback)(void));
 
+/** Callback function pointer for control transfer completion
+ * @param req Control request data
+ */
 typedef void (*usbd_control_complete_callback)(usbd_device *usbd_dev,
 		struct usb_setup_data *req);
 
+/** Callback function pointer for control transfers
+ * @param req Control request data,
+ * @param buf Buffer for DATA stage of control transfer, if applicable.
+ * @param len Size of buf
+ * @param complete Function to be called on completion.
+ * @return TODO : should return enum usbd_request_return_codes
+ */
 typedef int (*usbd_control_callback)(usbd_device *usbd_dev,
 		struct usb_setup_data *req, uint8_t **buf, uint16_t *len,
 		usbd_control_complete_callback *complete);
 
+/** Callback function pointer for standard Set Configuration requests
+ * @param wValue new configuration value
+ */
 typedef void (*usbd_set_config_callback)(usbd_device *usbd_dev, uint16_t wValue);
 
+/** Callback function pointer for standard Set Interface (alternate setting) requests
+ * @param wIndex Interface number
+ * @param wValue Alternate setting
+ */
 typedef void (*usbd_set_altsetting_callback)(usbd_device *usbd_dev,
 		uint16_t wIndex, uint16_t wValue);
 
+/** Callback function pointer for endpoint CTR event (?)
+ *
+ * Called when an OUT, IN, or SETUP token is received (?)
```

Comment:
I wasn't too sure about this. It seems up to each driver to decide when to call an usbd_endpoint_callback; they all appear to do so from within their _poll() function according to the type of token received. A second opinion would be required to validate this, though.